utilities/include/vec4.hpp

```diff
@@ -0,0 +1,149 @@
+
+        constexpr T a() const{
+            return m_v[3];
+        }
+        
+
+        constexpr T dot(const vec4& v2,const bool alpha=ALPHA_) const{
+            return m_v[0]*v2.m_v[0]+m_v[1]*v2.m_v[1]+m_v[2]*v2.m_v[2]+(m_v[3]*v2.m_v[3])*alpha;
+        }
+        
+        constexpr T norm(const bool alpha=ALPHA_) const{ //__GNUC__
+            return sqrt(dot(*this,alpha));
+        }
+
+        constexpr static T angle(const vec4& v1,const vec4& v2,const bool alpha=ALPHA_){//__GNUC__
+            return acos(v1.dot(v2,alpha)/(v1.norm(alpha)*v2.norm(alpha)));
+        }
+        constexpr vec4 unit() const{//__GNUC__
+            return *this/this->norm();
+        }
+
+        
+        constexpr vec4 cross(const vec4& v2) const{ //ignoring the "alpha" channel
+            return vec4(m_v[1] * v2.m_v[2] - m_v[2] * v2.m_v[1], m_v[2] * v2.m_v[0] - m_v[0] * v2.m_v[2],m_v[0] * v2.m_v[1] - m_v[1] * v2.m_v[0],0);
+        }
+
+        //vector addition
+        friend constexpr vec4 operator+(const vec4& v1, const vec4& v2){
+            return b2v(v2b(v1)+v2b(v2));
+        }
+
+        //vector subtraction
+        friend constexpr vec4 operator-(const vec4& v1, const vec4& v2){
+            return b2v(v2b(v1)-v2b(v2));
+        }
+
+        //scalar post-multiplication
+        friend constexpr vec4 operator*(const vec4& v1, const T& s){
+            return b2v(v2b(v1)*xsimd::batch<T,A>::broadcast(s));
+        }
+
+        //scalar pre-multiplication
+        friend constexpr vec4 operator*(const T& s, const vec4& v1){
+            return v1*s;
+        }
+
+        //scalar post-division
+        friend constexpr vec4 operator/(const vec4& v1, const T& s){
+            return v1*(1/double(s));
+        }
+
+        //addition+assignemnt
+        friend constexpr vec4& operator+=(vec4& v1, const vec4& v2){
+            v1=v1+v2;
+            return v1;
+        }
+        
+        //subtraction+assignemnt
+        friend constexpr vec4& operator-=(vec4& v1, const vec4& v2){
+            v1=v1-v2;
+            return v1;
+        }
+    
+        //scalar multiplication+assignment
+        friend constexpr vec4& operator*=(vec4& v1, const float& s){
+            v1=v1*s;
+            return v1;
+        }
+        
+        //scalar division+assignment
+        friend constexpr vec4& operator/=(vec4& v1, const float& s){
+            v1=v1/s;
+            return v1;
+        }
+
+        //negation operator 
+        friend constexpr vec4 operator-(const vec4& v1){
+            return b2v(-v2b(v1));
+        }
+
+        //indexing operation
+        constexpr T operator[](int i){
+            return m_v[i];
+        }
+        
+        //print the vector
+        friend std::ostream& operator<<(std::ostream& out, const vec4& v) {
+			out <<"("<< v.m_v[0] << "," << v.m_v[1] << "," << v.m_v[2]<<"," << v.m_v[3]<<")";
+			return out;
+		}
+
+	private:
+        T m_v[4];
+
+        static constexpr xsimd::batch<T,A> v2b(const vec4& v){
```

Comment:
Maybe a slightly more descriptive function name would be nicer.

Reply:
I didn't bother about it since it is not a part of API. Would a description of the function in the comments work fine?

Reply:
Regardless, having a difficult to understand function name makes it more difficult to understand what is happening in case we later need to make a change. For this reason, all function names should have descriptive names.

Reply:
will do.